### DISCUSSION TITLE
How to handle errors differently based on the URL?

### DISCUSSION THREAD

**Discussion by danechitoaie** [poem-web/poem#675] [Q&A] [answered]
I'm using poem-openapi and have mounted my API at `/api`.
```rust
    let app = Route::new()
        .at("/", get(index))
        .nest("/api", api_service)
        .nest("/doc", api_docs)
        .at("/spec.json", api_spec);
```

Is there any way to handle the errors differently based on URL path?
I would like that all requests going to `/api/*` to respond with JSON, so if there's an error (404, 400, 500, etc.) I want to be able to respond with something like `{"status": "error", "message": "..."}`.

And for any other URLs to respond with HTML output, so if there's an error (404, 400, 500, etc.) i want for this other non `/api/*` paths to respond with a HTML page showing that it was a 404, 400, 500, etc.

**Comment by NexRX** [comment, marked as answer]
You could use middleware to handle it outside your Endpoint like so

```rust
#[poem::async_trait]
impl<E: Endpoint> Endpoint for HandleEndpointErrorsMiddleware<E> {
    type Output = Response;

    async fn call(&self, req: Request) -> poem::Result<Self::Output> {
        trace!("Entered cache controlling middleware");
        let is_not_idempotent = req.method().is_safe();
        let mut res = self.ep.get_response(req).await;

        if res.status().is_client_error() || res.status().is_server_error() >= 300 { // or just res.status().as_u16() >= 300
            if req.original_uri().path().starts_with("/api/"){
                // Give Json errors
            } else {
                //Give HTML errors
            }
        }
    }
}

```

But I would actually personally recommend handling in the route/endpoint itself. If you want to abstract it away so it doesn't clutter the routes code then maybe you'd want a Uniform error type that implements `Into<Json>` or `Into<Html>` (or `From<UniformError> for Json`... etc), helps keep code consise where you need to.
